python/kotekan/shared_memory_buffer.py

```diff
@@ -0,0 +1,539 @@
+        logger.debug(
+            "Found {} time slots that are newer than {} in {}: {}.".format(
+                len(idxs), timestamp, last_ts, idxs
+            )
+        )
+
+        # select last n time slots
+        return VisRaw.from_nparray(
+            self._data[idxs, :],
+            self.size_frame,
+            len(idxs),
+            self.num_freq,
+            num_elements=7,
+            num_stack=28,
+            num_ev=0,
+        )
+
+    def _copy_from_shm(self, times, access_record):
+        # copy data updates within the last n time slots
+        for t in times:
+            for f_i in range(self.num_freq):
+                # check if this value should be copied: only if the access record changed and
+                # is not set to invalid
+                if access_record[t][f_i] != self.invalid_value and (
+                    self._last_access_record is None
+                    or access_record[t][f_i] > self._last_access_record[t][f_i]
+                ):
+                    logger.debug("Copying value at time={}, freq={}".format(t, f_i))
+                    # check if this time slot is in local copy of data
+                    try:
+                        t_i = self._time_index_map[access_record[t][f_i]]
+                    except KeyError:
+                        if len(self._time_index_map) == self.buffer_size:
+                            t_i = self._remove_oldest_time_slot()
+                        else:
+                            t_i = self._find_free_time_slot()
+                            if t_i is None:
+                                raise SharedMemoryError(
+                                    "Can't find a free slot in buffer (has keys {}).".format(
+                                        self._time_index_map.keys()
+                                    )
+                                )
+                        logger.debug(
+                            "Setting time index map [{}] to {}.".format(
+                                access_record[t][f_i], t_i
+                            )
+                        )
+                        self._time_index_map[access_record[t][f_i]] = t_i
+
+                    # copy the value
+                    # TODO: eliminate extra copy
+                    tmp = np.ndarray(
+                        1,
+                        np.dtype((np.void, self.size_frame)),
+                        self.shared_mem,
+                        self.pos_data + (t * self.num_freq + f_i) * self.size_frame,
+                        order="C",
+                    )
+                    self._data[t_i, f_i] = tmp[:]
+
+    def _access_record(self):
+        with self.semaphore:
+            record = np.ndarray(
+                (self.num_time, self.num_freq),
+                np.int64,
+                self.shared_mem,
+                self.pos_access_record,
+                order="C",
+            )
```

Comment:
I think you need to explicitly make a copy in here while holding the semaphore. I'm pretty sure this just creates a view of the underlying memory.